### PR TITLE
feat: add basic THORChain LP user data fetching utils

### PR DESCRIPTION
**Pull request by gomesalexandre**
## Description

Baby steps of THORCHain positions data fetching. Currently handles asymmetric positions on the asset side e.g deposing ETH only. Does *not* handle asym on the RUNE side, nor does it handle a *list* of positions for the same pool, or sym. deposits. To be added in a follow-up with units and fixtures.

- [x] test asym on the asset side is happy

The following will be brought in follow-up PRs. This is quite tricky to get right, so authoring the first bits so we can be unblocked on the view-layer side.

- [ ] ~~handle multiple positions for the same pool (sym on both sides + asym) - this needs to be done in conjunction with testin asym., as we will need to ensure the shape of the data we're getting in such case, they may use different identifiers~~ 
- [ ] ~~test asym on the RUNE side is happy~~
- [ ] ~~test sym. is happy~~
- [ ] ~~dedupes?~~

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

- contributes to https://github.com/shapeshift/web/issues/5939

## Risk

None, not consumed

### Engineering

- Deposit asymetrically on the asset side, and ensure data looks sane against THOR, though the shape of this is most likely to change in the direct follow-up

## Screenshots (if applicable)

Tested with a 0.01 ETH deposit asym (i.e 0.005 ETH and prorated RUNE)

- Parsed data

image

- THORYield

image

### CODE REVIEW

**Comment by gomesalexandre** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/shapeshift/web/5969?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

Current dependencies on/for this PR:
* `develop`
  * **PR #5966** <a href="https://app.graphite.dev/github/pr/shapeshift/web/5966?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #5969** <a href="https://app.graphite.dev/github/pr/shapeshift/web/5969?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/shapeshift/web/5969?utm_source=stack-comment).

File: src/lib/utils/thorchain/lp.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+}
+
+export const getThorchainLiquidityProviderPosition = async ({
+  accountId,
+  assetId,
+}: {
+  accountId: AccountId
+  assetId: AssetId
+}): Promise<LiquidityProvider | null> => {
+  const lendingPositionsResponse = await getAllThorchainLiquidityProviderPositions(assetId)
```
[the remark this answers is not in the document]
Comment (gomesalexandre) [reply src/lib/utils/thorchain/lp.ts:36]: 3bfe4175ec14d9b3071c062963e8934a6875136b

File: src/lib/utils/thorchain/lending.ts [hunk 2]
```diff
@@ -139,6 +139,8 @@ export const getThorchainLendingPosition = async ({
   accountId: AccountId
   assetId: AssetId
 }): Promise<Borrower | null> => {
+  // TODO(gomes): we want to use the /borrower/<address> endpoint for any chain other than UTXOs
+  // this is a big response, and can take a long time to be fetched
```
Comment (gomesalexandre) [inline src/lib/utils/thorchain/lending.ts:143]: Tackle me across the board before opening, might as well

File: src/pages/ThorChainLP/PoolsPage.tsx [hunk 3]
```diff
@@ -1,3 +1,16 @@
...
+
+import { useUserLpData } from './hooks/useUserLpData'
+
 export const PoolsPage = () => {
+  // TODO(gomes): for development only - remove me
+  const defaultAccountId = useAppSelector(state => selectFirstAccountIdByChainId(state, ethChainId))
+
+  const { data } = useUserLpData({ accountId: defaultAccountId, assetId: ethAssetId })
+
+  console.log({ data })
```
Comment (gomesalexandre) [inline src/pages/ThorChainLP/PoolsPage.tsx:13]: Not a forgotten console log - until we wire-up in a follow-up, here's our data!

File: src/pages/ThorChainLP/hooks/useUserLpData.ts [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+import { useAppSelector } from 'state/store'
+
+type UseLendingPositionDataProps = {
+  accountId: AccountId | undefined
+  assetId: AssetId
+}
+
+const calculatePoolOwnershipPercentage = (
+  userLiquidityUnits: string,
+  totalPoolUnits: string,
```
Comment (gomesalexandre) [inline src/pages/ThorChainLP/hooks/useUserLpData.ts:22]: LSP destructure not to get these wrong

File: src/pages/ThorChainLP/hooks/useUserLpData.ts [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+  queryKey: [string, { accountId: AccountId; assetId: AssetId }]
+}) => {
+  const [, { accountId, assetId }] = queryKey
+  const position = await getThorchainLendingPosition({ accountId, assetId })
+  return position
+}
+
+export const useUserLpData = ({ accountId, assetId }: UseLendingPositionDataProps) => {
+  // TODO(gomes): return a list of positions, and discriminate asym/sim as two different positions
+  // TODO(gomes): handle symmetric LP positions - this work was started with an asymmetric position with underlying rebalancing
```
Comment (gomesalexandre) [inline src/pages/ThorChainLP/hooks/useUserLpData.ts:37]: This should most likely work ootb but will need to be tested by depositing symetrically

File: src/pages/ThorChainLP/hooks/useUserLpData.ts [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+}: {
+  queryKey: [string, { accountId: AccountId; assetId: AssetId }]
+}) => {
+  const [, { accountId, assetId }] = queryKey
+  const position = await getThorchainLendingPosition({ accountId, assetId })
+  return position
+}
+
+export const useUserLpData = ({ accountId, assetId }: UseLendingPositionDataProps) => {
+  // TODO(gomes): return a list of positions, and discriminate asym/sim as two different positions
```
Comment (gomesalexandre) [inline src/pages/ThorChainLP/hooks/useUserLpData.ts:36]: Tackle me before opening - we absolutely want to handle this as early as possible

File: src/pages/ThorChainLP/hooks/useUserLpData.ts [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+      const underlyingAssetValueFiatUserCurrency = fromThorBaseUnit(
+        position?.asset_deposit_value || '0',
+      ).times(poolAssetMarketData?.price || 0)
+      const underlyingRuneValueFiatUserCurrency = fromThorBaseUnit(
+        position?.rune_deposit_value || '0',
+      ).times(runeMarketData?.price || 0)
+
+      // When depositing asymetrically, the assetDeposit value is the total amount deposited, but the asset_deposit_value is only half of that
+      // because of the underlying rebalancing
+      const isAsymmetric = position.assetDeposit !== position.asset_deposit_value
```
Comment (gomesalexandre) [inline src/pages/ThorChainLP/hooks/useUserLpData.ts:78]: Handle asym. on the RUNE side as well

File: src/pages/ThorChainLP/hooks/useUserLpData.ts (hunk 6 above)
Comment (gomesalexandre) [reply src/pages/ThorChainLP/hooks/useUserLpData.ts:36]: Actually not going to be tackled just yet but in a follow-up - getting the asym/sym right is quite complex, better to start with one happy case (asym on the asset side) and iterate from there

File: src/lib/utils/thorchain/lending.ts (hunk 2 above)
Comment (gomesalexandre) [reply src/lib/utils/thorchain/lending.ts]: b27a6c1c7b7069013edc6349001bd779c7159b8c

File: src/pages/ThorChainLP/hooks/useUserLpData.ts (hunk 4 above)
Comment (gomesalexandre) [reply src/pages/ThorChainLP/hooks/useUserLpData.ts]: a1c20f3b3c

File: src/pages/ThorChainLP/hooks/useUserLpData.ts (hunk 7 above)
Comment (gomesalexandre) [reply src/pages/ThorChainLP/hooks/useUserLpData.ts]: f00aa971db

File: src/pages/ThorChainLP/hooks/useUserLpData.ts [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+        .toFixed()
+
+      const poolOwnershipPercentage = calculatePoolOwnershipPercentage({
+        userLiquidityUnits: position.liquidityUnits,
+        totalPoolUnits: poolData.pool_units,
+      })
+
+      return {
+        underlyingAssetAmountCryptoPrecision: fromThorBaseUnit(
+          position.asset_deposit_value,
```
Comment (gomesalexandre) [inline src/pages/ThorChainLP/hooks/useUserLpData.ts:94]: Note, THOR only returns us the value at the time of deposit, i.e when depositing 0.01 ETH, 0.005 ETH and matching 0.23506872 RUNE.
To get the rebalanced value, we will need to calculate manually, see THORYield

image

Notice how there's now more underlying RUNE and less underlying ETH.

File: src/pages/ThorChainLP/hooks/useUserLpData.ts [hunk 9]
```diff
@@ -0,0 +1,111 @@
+import { type AccountId, type AssetId, thorchainAssetId } from '@shapeshiftoss/caip'
+import { useQuery } from '@tanstack/react-query'
+import axios from 'axios'
+import { getConfig } from 'config'
+import { useMemo } from 'react'
+import { bn, bnOrZero } from 'lib/bignumber/bignumber'
+import type { ThornodePoolResponse } from 'lib/swapper/swappers/ThorchainSwapper/types'
+import { assetIdToPoolAssetId } from 'lib/swapper/swappers/ThorchainSwapper/utils/poolAssetHelpers/poolAssetHelpers'
+import { fromThorBaseUnit } from 'lib/utils/thorchain'
+import { getThorchainLendingPosition } from 'lib/utils/thorchain/lending'
```
Comment (gomesalexandre) [inline src/pages/ThorChainLP/hooks/useUserLpData.ts:10]: Remove me

File: src/pages/ThorChainLP/hooks/useUserLpData.ts [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+export const thorchainLendingPositionQueryFn = async ({
+  queryKey,
+}: {
+  queryKey: [string, { accountId: AccountId; assetId: AssetId }]
+}) => {
+  const [, { accountId, assetId }] = queryKey
+  const position = await getThorchainLendingPosition({ accountId, assetId })
+  return position
+}
+
```
Comment (gomesalexandre) [inline src/pages/ThorChainLP/hooks/useUserLpData.ts:37]: Remove me

File: src/lib/utils/thorchain/lp.ts [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+
+    const accountAddresses = await getAccountAddresses(accountId)
+
+    return allPositions.find(position => accountAddresses.includes(position?.asset_address ?? ''))
+  })()
+  if (!accountPosition) return null
+
+  // TODO(gomes): asset_address *or* rune_address when implementing sim. pools
+  const { data: midgardLiquidityProvider } = await axios.get<MidgardLiquidityProvider>(
+    `${getConfig().REACT_APP_MIDGARD_URL}/member/${accountPosition.asset_address}`,
```
Comment (0xApotheosis) [inline src/lib/utils/thorchain/lp.ts:69]: We'll need to checksum `asset_address`.

✅ https://indexer.thorchain.shapeshift.com/v2/member/0x32DBc9Cf9E8FbCebE1e0a2ecF05Ed86Ca3096Cb6 (manually constructed)

❌ https://indexer.thorchain.shapeshift.com/v2/member/0x32dbc9cf9e8fbcebe1e0a2ecf05ed86ca3096cb6 (this PR)

**Review by 0xApotheosis — COMMENTED** [review]
Looking good, just an issue with the `asset_address` checksum.

File: src/lib/utils/thorchain/lp.ts (hunk 11 above)
Comment (gomesalexandre) [reply src/lib/utils/thorchain/lp.ts:69]: That's an odd one!
Was testing https://indexer.thorchain.shapeshift.com/v2/member/0x5daf465a9ccf64deb146eeae9e7bd40d6761c986 and it worked fine without chekcsum. In https://indexer.thorchain.shapeshift.com/v2/members your address isn't checksummed iether, but for some reason it looks like some addresses *may* require checksum so will implement it across the board

**Review by 0xApotheosis — APPROVED** [review]
Happy to get this one in as-is for velocity 👌

File: src/pages/ThorChainLP/hooks/useUserLpData.ts [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+      const underlyingAssetValueFiatUserCurrency = fromThorBaseUnit(
+        position?.asset_deposit_value || '0',
+      ).times(poolAssetMarketData?.price || 0)
+      const underlyingRuneValueFiatUserCurrency = fromThorBaseUnit(
+        position?.rune_deposit_value || '0',
+      ).times(runeMarketData?.price || 0)
+
+      const isAsymmetric =
+        bnOrZero(position.assetAdded).isZero() || bnOrZero(position.runeAdded).isZero()
+      const asymSide = bn(position.assetDeposit).gt(position.asset_deposit_value) ? 'asset' : 'rune'
```
[the remark this answers is not in the document]
Comment (gomesalexandre) [reply src/pages/ThorChainLP/hooks/useUserLpData.ts:68]: Done in the wire-up PR @0xApotheosis ! 0ed74d8e673b28ad970fa60691166d8d71f8c3bb

File: src/lib/utils/thorchain/lp.ts (hunk 11 above)
Comment (gomesalexandre) [reply src/lib/utils/thorchain/lp.ts:69]: Also tackled in the wire-up, gm! 68127fb39a9cd57e24fe8dc460fc2371527c1827
image

**Pull request merged**